faq/slurm.inc

```diff
@@ -91,15 +93,17 @@ href=\"?category=openfabrics#ib-locked-pages-more\">this FAQ entry for
 references about Slurm</a>.";
 
 /////////////////////////////////////////////////////////////////////////
+$q[] = "My job fails when using mpirun under Slurm 20.11";
 
-$q[] = "Any issues with Slurm 2.6.3?";
+$anchor[] = "slurm-20.11-mpirun";
 
-$anchor[] = "slurm-2.6.3-issue";
-
-$a[] = "Yes. The Slurm 2.6.3, 14.03 releases have a bug in their PMI-2
-support.
-
-For the slurm-2.6 branch, it is recommended to use the latest version
-(2.6.9 as of 2014/4), which is known to work properly with pmi2.
-
-For the slurm-14.03 branch, the fix will be in 14.03.1.";
+$a[] = "Due to the change in Slurm 20.11 - steps being exclusive by default
+using mpirun when build --with-slurm in sbatch script or under salloc may
+result in orted started process being limited to subset of resources. If you
+want to continue using mpirun in those places you should use [--whole] option
+with the built-in [srun] call. This can be achieved by the export of
+SLURM_WHOLE environment variable prior to mpirun execution:
+<geshi bash>
+export SLURM_WHOLE=1
+mpirun my_mpi_application
+</geshi>
```

Comment:
Might I suggest new wording (this is typed in github directly and not tested via PHP -- and note that the original PR is incorrect because it doesn't terminate the PHP string or line), and also having a 2nd FAQ item for jobs that are launched via `srun`, because those can be affected by this change, too ☹️ :

```php
$a[] = "SLURM 20.11 changed its default behavior that will affect [mpirun]'s behavior in all versions of Open MPI prior to v4.0.6.

When you use [mpirun] to launch an Open MPI application inside of a SLURM job (e.g., inside of an [salloc] or [sbatch]), [mpirun] actually uses [srun] under the covers to launch helper daemons on the nodes in your SLURM allocation.  Those helper daemons are then used to launch the individual MPI processes.

Starting with SLURM 20.11, by default, [srun] will associate a single Linux virtual processor with each SLURM task.  

More concretely: the SLURM daemon on each node in the SLURM allocation will create a Linux cgroup containing a single Linux virtual processor, and will launch the Open MPI helper daemon into that cgroup.  Consequently, the Open MPI helper daemon _and all MPI processes that are subsequently launched by that Open MPI helper daemon_ will be restricted to running on the single Linux virtual processor contained in the SLURM-created cgroup.

Put simply: in a given SLURM job, the Open MPI helper daemon and all MPI processes on the same node will be restricted to running on a single core (or hardware thread).

Starting with Open MPI v4.0.6, [mpirun] will automatically set the environment variable [SLURM_WHOLE] to [1] before invoking [srun], which will tell SLURM to return to its prior behavior of creating a cgroup that contains the all the Linux virtual processors that were allocated to the SLURM job.  This allows the Open MPI helper daemon and all of the MPI processes to spread out across the cores / hardware threads that were allocated to the SLURM job (vs. clumping them all together on a single core / hardware thread).

If you are using a version of Open MPI before v4.0.6, you should set the `SLURM_WHOLE` environment variable to [1] before invoking [mpirun].  For example:

<geshi bash>
shell$ export SLURM_WHOLE=1
shell$ mpirun my_mpi_application
</geshi>";

$q[] = "My job fails / performs poorly when using srun under Slurm 20.11 (and later)";

$anchor[] = "slurm-20.11-srun";

$a[] = "Similar to <a href=\"#slurm-20.11-mpirun\">the SLURM 20.11 issue with [mpirun]</a>, applications that use [srun] to directly launch Open MPI applications may experience a change in behavior compared to prior versions of SLURM.

Starting with SLURM 20.11, by default, [srun] will associate a single Linux virtual processor with each SLURM task.  

<ul>
<li> Prior to SLURM 20.11: by default, SLURM launches each MPI process in a cgroup that spans all the Linux virtual processors allocated to the SLURM job on that node.</li>
<li> Starting with SLURM 20.11: by default, SLURM launches each MPI process in a cgroup that contains a single Linux virtual processor (i.e., one of the Linux virtual processors allocated to the SLURM job on that node).</li>
</ul>

*NOTE:* Unlike when using [mpirun] in SLURM jobs, there are no helper Open MPI daemons involved in this process; [srun my_mpi_application] from inside a SLURM job will directly launch MPI processes on the nodes allocated to your SLURM job.

The consequence of this change is that if your MPI job requires more than one Linux virtual processor (i.e., more than one core and/or hardware thread), you will need to use additional CLI parameters (e.g., to [srun]) to tell SLURM to allocate additional resources to each task in your SLURM job.  For example, if you have a multi-threaded MPI application that benefits from utilizing multiple hardware cores and/or threads, you may need to tell [srun] to allocate more than one Linux virtual processor to each MPI process.

Consult the SLURM documentation for more details.";

Reply:
I feel that there is still more to the story here but being a FAQ I realize it also needs to be short. What I see is this:
`mpirun` from Open MPI 3.1.2 uses srun using a syntax as follows:
```
srun --ntasks-per-node=1 --kill-on-bad-exit --cpu_bind=none --nodes=<nsisters> --nodelist=<sister nodes> --ntasks=<nsisters> orted <various orted mca parameters>
```
the `cpu_bind` was later replaced by setting `SLURM_MPI_BIND=none` because `cpu_bind` was replaced by `cpu-bind` though this change was undone later in Slurm.

Now Slurm 20.11 in effect `--exclusive` to this by default. With an older Slurm you can see the effect of `--exclusive` quite clearly:

```
$ salloc --nodes=2 --tasks-per-node=2
...
$ srun --ntasks-per-node=1 --cpu_bind=none --nodes=1 --nodelist=sister --ntasks=1 numactl --show | grep phys
physcpubind: 2 6 
$ srun --ntasks-per-node=1 --cpu_bind=none --nodes=1 --nodelist=sister --ntasks=1 --exclusive numactl --show | grep phys
physcpubind: 2
```
i.e. you see with `--exclusive` that the sister is bound to a single core only now.
Then with Slurm 20.11 adding `--whole` makes it unbound (within the constraints of the overall allocation "jail") again.

(Now if users run srun manually we never tell them about `cpu-bind` but instead they need to use `--cpus-per-task` if they need one task bound to multiple CPUs.)